File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> sequence, int k) {

    vector<int> answer;
    
    int s_size = sequence.size();
    int sum = 0;
    int minVal = s_size;
    pair<int, int> candidate = {0, s_size - 1};

    int start = 0, end = 0;
    while (start < s_size) {
        if (sum == k && (end - start) < minVal) {
            candidate.first = start;
            candidate.second = end - 1;
            minVal = (end - start);
        } else if (sum < k && end < s_size) {
            sum += sequence[end++];
        } else {
            sum -= sequence[start++];
        }
    }

    answer.push_back(candidate.first);
    answer.push_back(candidate.second);
    return answer;
}

File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.java
import java.util.*;
    class Solution{
        public int[] solution(int[] nums, int target) {
            int[] answer = {};
            List<int[]> list = new ArrayList<>();
            int start = 0;
            int end = 0;
            int sum = nums[0];

            while (start <= end && end < nums.length) {
                if(sum == target){
                    list.add(new int[]{start, end});
                    sum -= nums[start];
                    start++;
                }
                else if(sum < target && end < nums.length-1){
                    end++;
                    sum += nums[end];
                }else{
                    sum -= nums[start];
                    start++;
                }
            }

            list.sort(new Comparator<int[]>() {
                @Override
                public int compare(int[] o1, int[] o2) {
                    return (o1[1] - o1[0]) - (o2[1] - o2[0]);
                }
            });
            answer = list.get(0);

            return answer;
        }
    }